oauthproxy.go

```diff
@@ -675,16 +693,32 @@ func (p *OAuthProxy) SignOut(rw http.ResponseWriter, req *http.Request) {
 // OAuthStart starts the OAuth2 authentication flow
 func (p *OAuthProxy) OAuthStart(rw http.ResponseWriter, req *http.Request) {
 	// start the flow permitting login URL query parameters to be overridden from the request URL
-	p.doOAuthStart(rw, req, req.URL.Query())
+
+	params := mux.Vars(req)
+	providerID := (params["id"])
+
+	if providerID == "" {
+		p.doOAuthStart(rw, req, p.defaultProvider, req.URL.Query())
+	} else {
+		if _, ok := p.providerMap[providerID]; ok {
+			p.doOAuthStart(rw, req, providerID, req.URL.Query())
+		} else {
+			logger.Errorf("Error while converting id %v from path: %v", providerID, req.URL.Path)
+			p.ErrorPage(rw, req, http.StatusNotFound, "Start Path does not match a provider ID")
+			return
+		}
+
+	}
 }
 
-func (p *OAuthProxy) doOAuthStart(rw http.ResponseWriter, req *http.Request, overrides url.Values) {
-	extraParams := p.provider.Data().LoginURLParams(overrides)
+func (p *OAuthProxy) doOAuthStart(rw http.ResponseWriter, req *http.Request, providerID string, overrides url.Values) {
+
+	extraParams := p.providerMap[providerID].Data().LoginURLParams(overrides)
```

Comment:
Perhaps we should fetch the provider out into a variable rather than fetching it from the map multiple times through this function